market/mk20/types.go

```diff
@@ -0,0 +1,244 @@
+	PieceCID cid.Cid `json:"piececid"`
+
+	// Size represents the size of the padded piece in the data source.
+	Size abi.PaddedPieceSize `json:"size"`
+
+	// Format defines the format of the piece data, which can include CAR, Aggregate, or Raw formats.
+	Format PieceDataFormat `json:"format"`
+
+	// SourceHTTP represents the HTTP-based source of piece data within a deal, including raw size and URLs for retrieval.
+	SourceHTTP *DataSourceHTTP `json:"sourcehttp"`
+
+	// SourceAggregate represents an aggregated source, comprising multiple data sources as pieces.
+	SourceAggregate *DataSourceAggregate `json:"sourceaggregate"`
+
+	// SourceOffline defines the data source for offline pieces, including raw size information.
+	SourceOffline *DataSourceOffline `json:"sourceoffline"`
+
+	// SourceHTTPPut // allow clients to push piece data after deal accepted, sort of like offline import
+	SourceHttpPut *DataSourceHttpPut `json:"sourcehttpput"`
+
+	// SourceStorageProvider -> sp IDs/ipni, pieceCids
+}
+
+// PieceDataFormat represents various formats in which piece data can be defined, including CAR files, aggregate formats, or raw byte data.
+type PieceDataFormat struct {
+
+	// Car represents the optional CAR file format, including its metadata and versioning details.
+	Car *FormatCar `json:"car"`
+
+	// Aggregate holds a reference to the aggregated format of piece data.
+	Aggregate *FormatAggregate `json:"aggregate"`
+
+	// Raw represents the raw format of the piece data, encapsulated as bytes.
+	Raw *FormatBytes `json:"raw"`
+}
+
+// FormatCar represents the CAR (Content Addressable aRchive) format with version metadata for piece data serialization.
+type FormatCar struct {
+	Version uint64 `json:"version"`
+}
+
+// FormatAggregate represents the aggregated format for piece data, identified by its type.
+type FormatAggregate struct {
+
+	// Type specifies the type of aggregation for data pieces, represented by an AggregateType value.
+	Type AggregateType `json:"type"`
+
+	// Sub holds a slice of PieceDataFormat, representing various formats of piece data aggregated under this format.
+	// The order must be same as segment index to avoid incorrect indexing of sub pieces in an aggregate
+	Sub []PieceDataFormat `json:"sub"`
+}
+
+// FormatBytes defines the raw byte representation of data as a format.
+type FormatBytes struct{}
+
+// DataSourceOffline represents the data source for offline pieces, including metadata such as the raw size of the piece.
+type DataSourceOffline struct {
+	RawSize uint64 `json:"rawsize"`
+}
+
+func (dso *DataSourceOffline) Name() DataSourceName {
+	return DataSourceNameOffline
+}
+
+func (dso *DataSourceOffline) IsEnabled(dbDataSources []dbDataSource) (int, error) {
+	name := string(dso.Name())
+	for _, p := range dbDataSources {
+		if p.Name == name {
+			if p.Enabled {
+				return Ok, nil
+			}
+		}
+	}
+	return ErrUnsupportedDataSource, xerrors.Errorf("data source %s is not enabled", name)
+}
```

Comment:
Imho this method shouldn't really exist, instead we should have a `Dependencies() map[Name]struct{}` which returns self + in case of aggregate the sub types.